metrics: Fix dtest->ulong conversion error

bhalevy opened the pull request:
prometheus: to_str(metric_value): fix conversion type for COUNTER and DERIVE
    
It looks like they got reversed.
    
Be consistent with metrics::impl::data_type documentation:
```
    enum class data_type : uint8_t {
        COUNTER, // unsigned int 64
        GAUGE, // double
        DERIVE, // signed int 64
        ABSOLUTE, // unsigned int 64
        HISTOGRAM,
    };
```
    
Fixes #899 

Test: heat_weighted_load_balancing_test.py:HeatWeightedLB.heat_weighted_load_balancing_cl_TWO_test(debug)
Verified that node logs show no ubsan errors not the added conversion warning.

amnonh commented:
Is this inline with collectd?

bhalevy commented:
Isn't `write_text_representation` called only for prometheus?

bhalevy commented:
> Is this inline with collectd?

Also, we're currently self-inconsistent with regards to handling the data type.
Are collectd data types in disagreement with prometheus' types?

amnonh commented:
> Isn't `write_text_representation` called only for prometheus?

It is, I didn't look at the series yet, just wante to know that if you make a change to the metrics layer collectd still works

amnonh commented:
> > Is this inline with collectd?
> 
> Also, we're currently self-inconsistent with regards to handling the data type.
> Are collectd data types in disagreement with prometheus' types?

Their not exactly the same, that's why I went to double as a common denominator.

bhalevy commented:
@amnonh From what I can see, `get_collectd_value` is unaffected since it doesn't interpret the value.
scollectd `write_value` looks fishy though:
https://github.com/scylladb/seastar/blob/d240dac462cfa7287c225e386a43a12a68b9d87f/src/core/scollectd.cc#L208-L225
as it calls `v.ui()` for `data_type::DERIVE` which seems wrong.
And not it may throw now if the value is negative.
I'll add a patch to fix that too.

bhalevy commented:
@amnonh ping

bhalevy commented:
@scylladb/seastar-maint please merge

nyh commented on src/core/prometheus.cc:
```diff
@@ -147,9 +147,9 @@ static std::string to_str(const seastar::metrics::impl::metric_value& v) {
-    case seastar::metrics::impl::data_type::DERIVE:
         return std::to_string(v.ui());
+    case seastar::metrics::impl::data_type::DERIVE:
+        return std::to_string(v.i());
```
You are right. The difference between COUNTER and DERIVE in collectd was already documented long ago in Scylla's docs/metrics.md, but it appears the implementation never matched the explanation. In any case, in Scylla, all COUNTER and DERIVE are positive, so this different didn't matter.

nyh reviewed (changes requested):
Looks mostly good, I just left a couple of minor questions/requests.

bhalevy commented on include/seastar/core/metrics.hh:
```diff
@@ -275,7 +277,12 @@ struct metric_value {
+        if (d < 0) {
+            ulong_conversion_error(d);
+            return 0;
+        }
```
good idea. I'll add it

nyh replied:
By the way, the lround(3) manual page says: "If x is a NaN or an infinity, or the rounded value is too large to be stored in a long ... the return value is unspecified."
So maybe the best thing is to have something like 
```
if (d >= 0 && d <= maxint) {
   // we're ok
   ...
} else {
   ulong_conversion_error()
}
```
The thinking in reversing the condition is that NaN < 0 won't be true, but this still needs to be a ulong_conversion_error.

bhalevy replied:
It's scary that !(d >= 0 && d <= max) is different then (d < 0 || d > max) with regards to NaN.

bhalevy commented on src/core/prometheus.cc:
```diff
@@ -603,7 +604,18 @@ future<> write_text_representation(output_stream<char>& out, const config& ctx,
+                    try {
+                        value_str = to_str(value);
+                    } catch (const std::range_error& e) {
+                        seastar_logger.warn("prometheus: write_text_representation: {}: {}", s.str(), e.what());
```
This happened due to to_str erroneously calling `ui()` for `DERIVE` rather than `i()`.
I agree it's not something that users can do anything about.
It would be helpful more for debugging purposes so we can indeed print it in debug level.

bhalevy commented on src/core/prometheus.cc:
```diff
@@ -603,7 +604,18 @@ future<> write_text_representation(output_stream<char>& out, const config& ctx,
+                        seastar_logger.warn("prometheus: write_text_representation: {}: {}", s.str(), e.what());
+                        value_str = "NaN";
+                    } catch (...) {
+                        auto ex = std::current_exception();
```
I think this error gets lost in `connection::start_response()` where `write_reply_to_connection` returns an exceptional future and `f.ignore_ready_future()` is called to ignore the exception.
The exception is not propagated to the real use via http.

bhalevy commented:
In v2:
- check `d >= 0 && d <= double(std::numeric_limits<long>::max())` for the valid bounds for `ui()` conversion
  - added similar check for `d >= min() && d <= max()` for the signed `i()` conversion.
- warning on range_error from to_str downgraded to debug log
- added comment for logging an error on any other exception

bhalevy commented:
@nyh ping
